src/abstractdataframe/abstractdataframe.jl

```diff
@@ -244,7 +244,11 @@ that is different than the number of rows present in `df`.
 """
 function Base.similar(df::AbstractDataFrame, rows::Integer = size(df, 1))
     rows < 0 && throw(ArgumentError("the number of rows must be non-negative"))
-    DataFrame(AbstractVector[similar(x, rows) for x in columns(df)], copy(index(df)))
```

Comment:
I'm not really convinced by the move from a comprehension to a longer loop. Am I right it only reduces the time taken by `similar` by 2%?

Reply:
I was getting various performance results that I did not understand (i.e. if the compiler worked correctly there should be no cost to this and I would expect that in the long run there should be 0% difference when Julia compiler matures).

The cost was up to around 10% that was originally indicated by @oxinabox, but most of the time it was several percent. I would be OK with leaving the comprehension, but I just tried to squeeze out something as this was the original reason for this PR (but I would keep the improved `getindex` methods even if we revert `similar` as they are touching the fundamental functionality of the package that is often used).

Reply:
my 10% was for `similar(df, 0)`
(which was the particular case that was worth optimising to me at the time I started looking at the code).

Reply:
Even 10% doesn't sound a lot. What kind of operation are you doing so that 10% of something as specific as `similar(df, 0)` matters?!

Reply:
In short, database query stuff. 
`simiilar(df, 0)` is what is made to make distinct empty results when nothing is returned.

Reply:
Sure, it's always worth investigating. But given that we don't save a lot, I don't think it's worth making the code more complex.

BTW, if  `similar(df, 0)` matters for your user case, you could probably create a special `AbstractDataFrame` object which would just store the column types, it should be dramatically faster.